### DISCUSSION TITLE
Intelligent pause of amended branch updates

### DISCUSSION THREAD

**Discussion by J4Numbers** [renovatebot/renovate#23823] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

As an overview, when I edit a RenovateBot branch by `--amend`-ing the RenovateBot commit to perform bug fixes, then I would expect it to be treated the same as appending new commits onto the branch and RenovateBot not rebasing on that branch any further.

### Background

Recently, when working on updating an outdated version of Terraform, I started from an already-created RenovateBot branch that bumped the version, then amended my own changes on top of that branch due to an organisational preference for single-commit branches.

At a certain point in time, the `develop` branch was updated, and the RenovateBot branch was rebased on top of that with only the Terraform version bump, and with none of the additional fixes that were performed.

I did have the changes saved locally, so I was able to force the change back again, but the previously mentioned existing behaviour for updated branches makes this feel like a bug.

### Reproduction repo

> Because I don't control my organisation's usage of RenovateBot (and a lot of the repos are hidden behind policies), I've reproduced the above on GitHub using the RenovateBot dashboard on Mend. The full PR history is available at https://github.com/J4Numbers/renovate-amend-rebase/pull/3

1. Baseline commit branch raised by RenovateBot: https://github.com/J4Numbers/renovate-amend-rebase/commit/2813050b85df0dd3e4bea11415690e797b893fd6
1. Amended RenovateBot commit: https://github.com/J4Numbers/renovate-amend-rebase/commit/a68ba24f35818fb617d62efc7683c5dce38b13c5
2. Updated default branch with a nominal change: https://github.com/J4Numbers/renovate-amend-rebase/commit/c1bbb384eadecb23a50b32390dc5c49d499469c7
3. Re-ran RenovateBot manually and the branch was overwritten with the base commit: https://github.com/J4Numbers/renovate-amend-rebase/commit/59654654e4c4a8204e017ffa53691c5182e881a9

### Requested behaviour

**Given** a RenovateBot branch where an update has been performed  
**And** the squashed diff of those changes is the same as the base RenovateBot update  
**When** the default branch of the project is updated
**Then** the updated RenovateBot branch is rebased against the default branch

**Given** a RenovateBot branch where an update has been performed  
**And** the squashed diff of those changes is different to the base RenovateBot update  
**When** the default branch of the project is updated
**Then** the updated RenovateBot branch is not rebased against the default branch

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: 1 already existing branches found: renovate/chalk-5.x-lockfile
DEBUG: Branch concurrent limit remaining: 9
DEBUG: Calculated maximum branches remaining this run: 1
DEBUG: Branches limit = 1
DEBUG: syncBranchState() (branch="renovate/chalk-5.x-lockfile")
DEBUG: syncBranchState(): update baseBranchSha (branch="renovate/chalk-5.x-lockfile")
DEBUG: syncBranchState(): update branchSha (branch="renovate/chalk-5.x-lockfile")
DEBUG: getBranchPr(renovate/chalk-5.x-lockfile) (branch="renovate/chalk-5.x-lockfile")
DEBUG: findPr(renovate/chalk-5.x-lockfile, undefined, open) (branch="renovate/chalk-5.x-lockfile")
DEBUG: Found PR #3 (branch="renovate/chalk-5.x-lockfile")
DEBUG: branchExists=true (branch="renovate/chalk-5.x-lockfile")
DEBUG: dependencyDashboardCheck=undefined (branch="renovate/chalk-5.x-lockfile")
DEBUG: PR rebase requested=false (branch="renovate/chalk-5.x-lockfile")
DEBUG: Checking if PR has been edited (branch="renovate/chalk-5.x-lockfile")
DEBUG: branch.isModified(): using git to calculate (branch="renovate/chalk-5.x-lockfile")
DEBUG: branch.isModified() = false (branch="renovate/chalk-5.x-lockfile")
DEBUG: Found existing branch PR (branch="renovate/chalk-5.x-lockfile")
DEBUG: Checking schedule(at any time, null) (branch="renovate/chalk-5.x-lockfile")
DEBUG: No schedule defined (branch="renovate/chalk-5.x-lockfile")
DEBUG: Branch already exists (branch="renovate/chalk-5.x-lockfile")
DEBUG: GET https://api.github.com/repos/J4Numbers/renovate-amend-rebase/branches/main/protection = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=86) (branch="renovate/chalk-5.x-lockfile")
DEBUG: No branch protection found (branch="renovate/chalk-5.x-lockfile")
DEBUG: Skipping behind base branch check due to rebaseWhen=auto (branch="renovate/chalk-5.x-lockfile")
DEBUG: isBranchConflicted(main, renovate/chalk-5.x-lockfile) (branch="renovate/chalk-5.x-lockfile")
DEBUG: branch.isConflicted(): using git to calculate (branch="renovate/chalk-5.x-lockfile")
DEBUG: Setting git author name: renovate[bot] (branch="renovate/chalk-5.x-lockfile")
DEBUG: Setting git author email: 29139614+renovate[bot]@users.noreply.github.com (branch="renovate/chalk-5.x-lockfile")
DEBUG: branch.isConflicted(): true (branch="renovate/chalk-5.x-lockfile")
DEBUG: Branch is conflicted (branch="renovate/chalk-5.x-lockfile")
DEBUG: Branch is not mergeable and needs rebasing (branch="renovate/chalk-5.x-lockfile")
DEBUG: Using reuseExistingBranch: false (branch="renovate/chalk-5.x-lockfile")
DEBUG: Setting current branch to main (branch="renovate/chalk-5.x-lockfile")
DEBUG: latest commit (branch="renovate/chalk-5.x-lockfile")
{
  "branchName": "main"
  "latestCommitDate": "2023-08-11T09:31:40+01:00"
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by viceice** [comment, marked as answer]
not possible, so don't use amend. push separate commit instead. you can use squash merge to combine them later to one commit on target branch.

**Reply by HonkingGoose** [reply]
> As an overview, when I edit a RenovateBot branch by `--amend`-ing the RenovateBot commit to perform bug fixes, then I would expect it to be treated the same as appending new commits onto the branch and RenovateBot not rebasing on that branch any further.

Pushing an amended commit is _not the same_ as pushing a new commit. When you amend you usually do one of these things:

- Change the commit message
- Change the actual content of the commit

I guess Renovate doesn't check if the `sha1` of its own commit has changed, due to a `git commit --amend && git push`. I think Renovate only react to _new_ commits. The [Renovate docs, Updating and rebasing branches](https://docs.renovatebot.com/updating-rebasing/) says:

> ## No rebasing if you have made edits
> 
> First of all, here is the one time when Renovate _won't_ update branches.
> If you edit a Renovate branch directly (e.g. to make a code fix to allow tests to pass again) then Renovate stops all updates of that branch.
> It is up to you to either finish the job and merge the PR, or rename it and close it so that Renovate can take back over again.

I'll update our docs to make it clear that Renovate only "sees new commits", and ignores any amended commits.

---

Here's the PR:

- #23833

**Reply by viceice** [reply]
renovate only compares the commit author, so when you amend, renovate sees that the branch inst up to date with expected content and because the author is renovate instead of somebody else it will push a new commit.

**Reply by HonkingGoose** [reply]
Thanks for the explanation @viceice. 🙂
